keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1197 @@
+is an extension of DRA that addresses this by making claim parameters less
+opaque. It also enables the scheduler to allocate claims rapidly, without
+back-and-forth communication with DRA drivers.
+
+The underlying concept is that drivers manage resources. Each resource has a
+set of attributes which describe the resource and its capabilities, for example
+the ability to provide a certain programming API at a specific version. Each
+resource may have a set of resource dimensions which are requestable by
+workloads, for example a certain amount of RAM. A resource might also be a pool
+of instances with the same capabilities. Then the resource dimension is the
+amount of those instances.
+
+Information about available resources get published in NodeResourceSlice
+objects in the apiserver.
+
+When a user wants to consume resources, they create a ResourceClaim, which
+references a claim parameter object. That object defines how much resources are
+needed and which capabilities they must have. Typically that object is using a
+vendor provided type which might also supports configuration parameters that
+are not needed for resource allocation.
+
+A driver-specific process "resolves" that request into a canonical form (the
+ResourceClaimParameters type) which components such as a scheduler or
+autoscaler can evaluate against the NodeResourceSlices of candidate nodes,
+without knowing exactly what is being requested.
+
+When a workload starts, an on-node driver is responsible for making the
+requested device(s) active in the Pod, according to the resource choices made
+by the scheduler.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Enable cluster autoscaling when pods use resource claims, with correct
+  decisions and changing the cluster size by more than one node at a time.
+
+- Support node-local resources. Adding or removing nodes has no effect
+  on network-attached resources and therefore CA does not need to (and cannot)
+  simulate them.
+
+- Allow DRA driver developers to provide a user experience that is similar to
+  the one possible without semantic parameters. Ideally, users should not notice
+  at all that a driver is using semantic parameters under the hood.
+
+### Non-Goals
+
+- Support semantic class parameters. They remain opaque and can be used to
+  define configuration of resources, but must not affect where a claim can be
+  allocated.
+
+- Scheduling performance is expected to become better compared to using the
+  PodSchedulingContext. However, this is not the reason for this KEP.
+
+
+## Proposal
```

Comment:
Where are user stories?

I think we should summarize which kind of use cases are supported, like:
- devices with arbitrary subresources and metadata
- pool of cards

are we supporting node topology?  If not fully, are there simplified cases of topology that can be supported with the proposed models?

Reply:
The revised intro explains what is possible with the new semantic model.

We have thought about topology a bit and believe that topology constraints can be done based on that semantic model, but how to describe it specifically is still open and not defined yet. There are some comments about that further down.